### ISSUE TITLE
Stops reading after first rarfile

### ISSUE THREAD

**Issue by Meliox** [hasse69/rar2fs#45]
Hi,

I'm having a annoying problem, in which rar2fs stops reading den file after it has read the first rarfile part. So this means that only the first part is read. This could be 15MB. I did not have the problem prior updating to 1.22, yet I think the error is due to something in unrar. Anyways, here's the debug log if that can be helpful.

log: http://pastebin.com/SbVG480c

 rar2fs -V
rar2fs v1.22.0 (DLL version 7)    Copyright (C) 2009 Hans Beckerus
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it under
certain conditions; see <http://www.gnu.org/licenses/> for details.
FUSE library version: 2.9.2
fusermount version: 2.9.2
using FUSE kernel interface version 7.19

unrar:
#define RARVER_MAJOR     5
#define RARVER_MINOR    31
#define RARVER_BETA      0
#define RARVER_DAY       4
#define RARVER_MONTH     2
#define RARVER_YEAR   2016

**Comment by hasse69** [comment]
Thanks for the issue report.

From the first look of it it seems like a really basic function is broken. But I cannot seem to be able to reproduce it. I have tried with both classic .rXX archives and newer .partXXX variants.
Are you saying this happens for all your volumes or only some of them? 
I will test also using the exact same version of libunrar as you are using, but I would be very surprised if it did not work. The best would of course be if you could send me a pointer to an archive that fails for you somehow so I can test it from here too.

**Comment by Meliox** [comment]
Thanks for you reply.

It happens for all .rXX archives. I don't think I have any .partXXX variants.

Well, perhaps I compiled it in a wrong way. 

Either I can give you access through teamviewer + terminal or you can give me a clue to your email :)

**Comment by hasse69** [comment]
Ok. I have now tried the same version of libunrar as you have. I still do not see any issues.
Can you please provide details about what options you use when mounting. I assume the archives are in store format and not compressed?
Did you also replace libunrar.so in your system path with the version you pointed rar2fs at when building it? Try `ldd rar2fs` and make sure the library libunrar.so points is in fact the one you built.
Again I assume you also built libunrar.so and not only rar2fs.
Another thing we can try is to re-configure your package with full debug and start it with the -f flag, skip -d because it will give us nothing in this case.

`./configure <what-ever-options-you-use-now> --enable-debug`

Note that this will produce a lot of output.

**Comment by Meliox** [comment]
Yes, they are in store format.
I'm using this in fstab:
rar2fs#/ftp/ //ftp/NORAR fuse ro,allow_other,noatime,uid=ammin,gid=ammin,umask=222,--seek-length=2 0 0
And for producing the debug above I simply used rar2fs /ftp/ //ftp/NORAR -o allow_other

 ldd rar2fs
        linux-vdso.so.1 =>  (0x00007fff66590000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f8032d3c000)
        libunrar.so => /usr/lib/libunrar.so (0x00007f8032ae5000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f80328c7000)
        libfuse.so.2 => /lib/x86_64-linux-gnu/libfuse.so.2 (0x00007f803268c000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f8032476000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f80320b1000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8031dab000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f8033040000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f8031ba7000)

I did change the config.mk to this, before compiling. Otherwise no options were used.

# Change below to match current configuration
##########################

##########################
# This is the path (absolute or relative) to the folder containing
# the full portable "Unrar C++ library" (libunrar) sources
UNRAR_SRC=/home/ammin/src/unrar

##########################
# This is the path (absolute or relative) to the folder containing
# the compiled libunrar.so.
# This can be left blank if the system already points out the location
# of a compatible unrarlib.so, eg. /lib.
UNRAR_LIB=/home/ammin/src/unrar

I will try and re-conifgure tomorrow.to provide a full debug.

Thanks.

**Comment by hasse69** [comment]
Ouch! You are using the old and deprecated config.mk way of building this? 
I honestly do not know if that would work anymore. Been ages since I saw anyone use that.
I would recommend you use the autotools way of building instead.

But still, you did not really answer my question regarding libunrar.so.
Did you rebuild libunrar.so and did you then install it in /usr/lib/ using the same source as you pointed rar2fs to in config.mk?

Building using config.mk might work. I see not real reason it should not.
But to enable the requested debug level you need to use autotools I am afraid.

According to what I see in your config.mk it should be something in the line of:

`# ./configure --with-unrar=/home/ammin/src/unrar --with-unrar-lib=/home/ammin/src/unrar --enable--debug`
`# make`

**Comment by Meliox** [comment]
Yes... And now I see that was a stupid mistake. I removed the git clone, and cloned another. Then ran autotools, and configure "./configure --with-unrar=/home/ammin/src/unrar --with-unrar-lib=/home/ammin/src/unrar --enable-debug"; make; make install.

Now rar2fs can read all the parts and copy does not stop after first part. I'm unsure why the difference between config.mk and autotools should be that large. Well, I think I confirmed that it isn't working anymore.

Yes, I did copy the libunrar.so it seems :)

$ md5sum /usr/lib/libunrar.so
b78846af303926f21d0ab15af658c5fc  /usr/lib/libunrar.so
$ md5sum /home/ammin/src/unrar/libunrar.so
b78846af303926f21d0ab15af658c5fc  libunrar.so

**Comment by hasse69** [comment]
Ok, that is good news I guess.
I will have a look at the confik.mk build, because I cannot really understand either why it would make a difference really. I will keep this issue open but I will lower the priority since it seems that you are good to go now?

**Comment by Meliox** [comment]
Indeed. Thanks for leading me in the right direction, and yes I suppose it's low priority as it works now.

**Comment by hasse69** [comment]
Just for the sake of it, can you try to add -DRAR_SMP to the compile switches in config.mk?

**Comment by Meliox** [comment]
You need to specify the exact location..

ifndef CROSS
# Host/target platform using GCC
CC=gcc
CXX=g++
CFLAGS+=-g -fno-omit-frame-pointer
CXXFLAGS+=-g -fno-omit-frame-pointer
DEFINES=-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE
STRIP=strip
LDFLAGS=
else
# Cross-compile using GCC
CC=$(CROSS)-gcc
CXX=$(CROSS)-g++
CFLAGS+=-g
CXXFLAGS+=-g
DEFINES=-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE
STRIP=$(CROSS)-strip
LDFLAGS=
endif

**Comment by hasse69** [comment]
`DEFINES=-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DRAR_SMP`
You can change in both places but the former is the one you use unless you are cross-compiling.

**Comment by Meliox** [comment]
Adding -DRAR_SMP to the compile switches in config.mk also fixed to problem. Good find, however, I will use autotools from now on.

**Comment by hasse69** [comment]
Yes, I would advise to stop using config.mk since I do not actively check its functionality. But since it is still there I should fix problems when reported. The -DRAR_SMP file is a bit tricky since it was added as part of RAR5 and it is possible to build libunrar without it. But the build of rar2fs needs to match whatever setting was used. The autotools build is clever enough to figure out when and if this flag is needed or not. But the confik.mk has no such mechanism. Since libunrar use -DRAR_SMP as default that is probably what I need to add to config.mk too. But optimally removing config.mk completely is probably the long term solution. Problem is, I do not know if there is infact systems still using this because of lack of autotools support :(

**Comment by Meliox** [comment]
Well. Is it not possible to stop supporting config.mk from a specific version of rar2fs? Yet, I can see some systems like routers etc. that might not have autotools. Might be, that you could mention in the wiki that if libunrar is build without -DRAR_SMP, then you should remove that compile switch from config.mk. Not an optimal solution, yet might work.

**Comment by hasse69** [comment]
Just for fun, try to replace your Makefile.mk with this one (after removing the .txt extension) and try again. Remove any settings of -DRAR_SMP you did before in config.mk.
If all goes as planned it should now autodetect the need for -DRAR_SMP.

[Makefile.mk.txt](https://github.com/hasse69/rar2fs/files/134869/Makefile.mk.txt)

**Comment by Meliox** [comment]
Cannot get it working..

`make -f Makefile.mk
g++ -g -fno-omit-frame-pointer -O2 -rdynamic  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_GNU_SOURCE -DRARDLL   -MD -I. -I -c dllext.cpp
dllext.cpp:30:23: fatal error: version.hpp: No such file or directory
 #include "version.hpp"
                       ^
compilation terminated.
make: *** [dllext.o] Error 1`

**Comment by Meliox** [comment]
It works :)

**Comment by hasse69** [comment]
Excellent! Thanks a lot for your help in verifying this. A lot can be done locally but nothing beats a test subject :) I will clean up the file and push it to master. Once that is done I will close this issue.

**Comment by Meliox** [comment]
Happy to be of help. And once again thanks for helping me solving me issue :)

**Comment by hasse69** [comment]
Btw, by looking at your attached debug log I think you should benefit a lot from adding the following to your mount options: `--exclude=.ftpaccess`
Check out the man page for a detailed description of why that might be a good idea.

**Comment by Meliox** [comment]
Appreciate it. Could definitely use some speed improvements. Thanks.

2016-02-17 23:49 GMT+01:00 hasse69 <notifications@github.com>:
